src/python/pants/backend/shell/shunit2_test_runner.py

```diff
@@ -0,0 +1,244 @@
+    process: Process
+
+
+class Shunit2AlreadySourced(Exception):
+    pass
+
+
+class ShellNotConfigured(Exception):
+    pass
+
+
+def validate_source_shunit2_not_included(fc: FileContent, address: Address) -> None:
+    lines = fc.content.splitlines()
+    for i, line in enumerate(lines):
+        if re.search(rb"(?:source|\.) +[.${}/'\"\w]*shunit2\b", line):
+            raise Shunit2AlreadySourced(
+                f"The test file {fc.path} sources shunit2 on line {i + 1} with: "
+                f"{line.decode()}\n\n"
+                f"Please remove this line so that Pants can run shunit2 for the target {address}, "
+                "as Pants will download the shunit2 script and source it for you automatically."
+            )
+
+
+@dataclass(frozen=True)
+class Shunit2RunnerRequest:
+    address: Address
+    test_file_content: FileContent
+    shell_field: Shunit2ShellField
+
+
+@dataclass(frozen=True)
+class Shunit2Runner:
+    shell: Shunit2Shell
+    binary_path: BinaryPath
+
+
+@rule(desc="Determine shunit2 shell")
+async def determine_shunit2_shell(
+    request: Shunit2RunnerRequest, shell_setup: ShellSetup
+) -> Shunit2Runner:
+    if request.shell_field.value is not None:
+        tgt_shell = Shunit2Shell(request.shell_field.value)
+    else:
+        parse_result = Shunit2Shell.parse_shebang(request.test_file_content.content)
+        if parse_result is None:
+            raise ShellNotConfigured(
+                f"Could not determine which shell to use to run shunit2 on {request.address}.\n\n"
+                f"Please either specify the `{Shunit2ShellField.alias}` field or add a "
+                f"shebang to {request.test_file_content.path} with one of the supported shells in "
+                f"the format `!#/path/to/shell` or `!#/path/to/env shell`"
+                f"(run `./pants help {Shunit2Tests.alias}` for valid shells)."
+            )
+        tgt_shell = parse_result
+
+    env = await Get(Environment, EnvironmentRequest(["PATH"]))
+    path_request = BinaryPathRequest(
+        binary_name=tgt_shell.name,
+        search_path=shell_setup.executable_search_path(env),
+        test=tgt_shell.binary_path_test,
+    )
+    paths = await Get(BinaryPaths, BinaryPathRequest, path_request)
+    first_path = paths.first_path
+    if not first_path:
+        raise BinaryNotFoundError(path_request, rationale=f"run shunit2 on {request.address}")
+    return Shunit2Runner(tgt_shell, first_path)
+
+
+@rule(desc="Setup shunit2", level=LogLevel.DEBUG)
+async def setup_shunit2_for_target(
+    request: TestSetupRequest,
+    shell_setup: ShellSetup,
+    test_subsystem: TestSubsystem,
+    test_extra_env: TestExtraEnv,
+    global_options: GlobalOptions,
+) -> TestSetup:
+    shunit2_download_file = DownloadFile(
+        "https://raw.githubusercontent.com/kward/shunit2/b9102bb763cc603b3115ed30a5648bf950548097/shunit2",
+        FileDigest("1f11477b7948150d1ca50cdd41d89be4ed2acd137e26d2e0fe23966d0e272cc5", 40987),
```

Comment:
Do we want to expose this via the options system? It felt like overkill to me, esp. because the user would need to specify the URL, hash, and # bytes.